src/java.base/share/classes/jdk/internal/foreign/AbstractMemorySegmentImpl.java

```diff
@@ -487,8 +487,8 @@ public int characteristics() {
     @Override
     public String toString() {
         return "MemorySegment{ " +
-                heapBase().map(hb -> "heapBase: " + hb + ", ").orElse("") +
-                "address: " + Utils.toHexString(address()) +
+                heapBase().map(hb -> "heapBase: " + hb).orElse(isMapped() ? "mapped" : "native") +
```

Comment:
This is not correct. Heap base can be empty for read-only segments.
Have we considered printing one of:
* HeapMemorySegment
* NativeMemorySegment
* MappedMemorySegment
E.g. add the heap/native/mapped as part of the "class" qualifier, before the `{` ?